File: internal/ui/clipboard.go
package ui

import (
	. "github.com/vipmax/edgo/internal/operations"
	. "github.com/vipmax/edgo/internal/utils"
	"github.com/atotto/clipboard"
)

// this function sets the required parameters to validate updates
func (e *Editor) set_update_parameters(changed bool) {
	e.Update = true
	e.IsContentChanged = changed
	e.FindTests()
}

func (e *Editor) OnCopy() {
	selectionString := e.Selection.GetSelectionString(e.code.Value())
	clipboard.WriteAll(selectionString)
}

func (e *Editor) OnPaste() {

	if e.Selection.IsSelectionNonEmpty() {
		e.Cut(false)
	}

	text, _ := clipboard.ReadAll()

	if len(text) == 0 { return }
	e.InsertString(e.Row, e.Col, text)
	e.set_update_parameters(true)
}

func (e *Editor) Cut(isCopySelected bool) {
	e.Focus()

	if (e.code.Len() < 0) { e.Row, e.Col = 0, 0; return }

	if isCopySelected {
		selectionString := e.Selection.GetSelectionString(e.code.Value())
		if len(selectionString) == 0 { return } // skipping function if selection is empty
		clipboard.WriteAll(selectionString)
	}

	selectedIndices := e.Selection.GetSelectedIndices(e.code.Value())
	lastElement := len(selectedIndices) - 1

	exd := selectedIndices[lastElement][0]
	eyd := selectedIndices[lastElement][1]
	sxd := selectedIndices[0][0]
	syd := selectedIndices[0][1]

	e.Row = syd
	e.Col = sxd

	sxd += LineOffset(e.code.Value(), syd) + 1
	exd += LineOffset(e.code.Value(), eyd) + 2

	text := make([]byte, exd - sxd)
	copy(text, e.code.Slice(sxd, exd))

	e.code.Remove(sxd, exd)
	e.treeSitterHighlighter.RemoveTextEdit(e.code.Value(), sxd, len(text))

	e.Undo = append(e.Undo, EditOperation{{Delete, text, sxd, CursorMove{e.Row, e.Col}}})
	e.Selection.CleanSelection()
	e.UpdateLsp(false, string(e.code.Value()))
	e.set_update_parameters(true)
}

func (e *Editor) Duplicate() {
	e.Focus()

	if e.code.Len() == 0 { return }
	syd := LineOffset(e.code.Value(), e.Row) + 1
	eyd := LineOffset(e.code.Value(), e.Row + 1) + 1
	if e.Row == 0 { syd-- } // this is required for first line only
	duplicatedSlice := e.code.Slice(syd, eyd)

	e.code.Insert(eyd, duplicatedSlice)
	e.Row++

	eyd = LineOffset(e.code.Value(), e.Row) + 1

	e.treeSitterHighlighter.InsertTextEdit(e.code.Value(), syd, len(duplicatedSlice))
	e.Undo = append(e.Undo, EditOperation{{Insert, duplicatedSlice, eyd, CursorMove{e.Row, e.Col}}})
	e.UpdateLsp(false, string(e.code.Value()))
	e.set_update_parameters(true)
}

func (e *Editor) OnUndo() {
	if len(e.Undo) == 0 { e.UpdateLsp(true, string(e.code.Value())); return }

	lastOperation := e.Undo[len(e.Undo)-1]
	e.Undo = e.Undo[:len(e.Undo)-1]
	e.Focus()

	index := len(lastOperation) - 1
	var o Operation
undo:
	if len(lastOperation) > 0 { o = lastOperation[index] } else { goto exit }

	switch o.Action {
	case Insert, AddChar:
		e.code.Remove(o.Offset, o.Offset + len(o.Text))
		e.treeSitterHighlighter.RemoveTextEdit(e.code.Value(), o.Offset, len(o.Text))
		e.Row = o.Cursor.Line; e.Col = o.Cursor.Column
	break;
	case Delete:
		e.code.Insert(o.Offset, o.Text)
		e.treeSitterHighlighter.InsertTextEdit(e.code.Value(), o.Offset, len(o.Text))
		e.Row = o.Cursor.Line; e.Col = o.Cursor.Column
	break;
	}

	if index > 0 { index--; goto undo; }
exit:
	e.Redo = append(e.Redo, lastOperation)
	e.UpdateLsp(false, string(e.code.Value()))
	e.set_update_parameters(true)
}

func (e *Editor) OnRedo() {
	if len(e.Redo) == 0 { return }

	lastRedoOperation := e.Redo[len(e.Redo)-1]
	e.Redo = e.Redo[:len(e.Redo)-1]

	index, offset := 0, 0
	var o Operation
redo:
	if len(lastRedoOperation) > 0 { o = lastRedoOperation[index] } else { goto exit }

	switch o.Action {
	case AddChar:
		offset++
		fallthrough
	case Insert:
		e.code.Insert(o.Offset, o.Text)
		e.treeSitterHighlighter.InsertTextEdit(e.code.Value(), o.Offset, len(o.Text))
		e.Row = o.Cursor.Line; e.Col = o.Cursor.Column + offset
	break;
	case Delete:
		e.code.Remove(o.Offset, o.Offset + len(o.Text))
		e.treeSitterHighlighter.RemoveTextEdit(e.code.Value(), o.Offset, len(o.Text))
		e.Row = o.Cursor.Line; e.Col = o.Cursor.Column
	break;
	}

	if index < len(lastRedoOperation) - 1 { index++; goto redo; }
exit:
	e.Undo = append(e.Undo, lastRedoOperation)
	e.UpdateLsp(false, string(e.code.Value()))
	e.set_update_parameters(true)
}

File: source/clipboard.cpp
/**
    Copyright (C) 2023 - 2026, edgo authors

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, see <https://www.gnu.org/licenses/>.
*/

#include "editor.h"
#include "operations.h"
#include "utils.h"
#include "sdl2_init.h"
#include "highlighter.h"

// this function sets the required parameters to validate updates
void set_update_parameters(bool changed) {
	e.update = true;
	e.isContentChanged = changed;
//	FindTests();
}

void OnCopy(void) {
	auto selectionString = e.__selection.GetSelectionString(e.code_str());
	__set_clipboard_text(selectionString.c_str());
}

void OnPaste(void) {
	if (e.__selection.IsSelectionNonEmpty()) {
		Cut(false);
	}

	char* text = __get_clipboard_text();

	if (strlen(text) == 0) { return; }
	InsertString(e.row, e.col, std::string(text));
	__free_clipboard(text);
	set_update_parameters(true);
}

void Cut(bool isCopySelected) {
	Focus();

	if (e.code.size() == 0) { e.row = 0; e.col = 0; return; }

	if (isCopySelected) {
		auto selectionString = e.__selection.GetSelectionString(e.code_str());
		if (selectionString.length() == 0) { return; } // skipping function if selection is empty
		__set_clipboard_text(selectionString.c_str());
	}

	auto selectedIndices = e.__selection.GetSelectedIndices(e.code_str());
	int lastElement = selectedIndices.size() - 1;

	auto exd = selectedIndices[lastElement].x;
	auto eyd = selectedIndices[lastElement].y;
	auto sxd = selectedIndices[0].x;
	auto syd = selectedIndices[0].y;

	e.row = syd;
	e.col = sxd;

	int exd_offset = 2;
	if (sxd == 0) { exd_offset++; }

	sxd += LineOffset(e.code_str(), syd) + 1;
	exd += LineOffset(e.code_str(), eyd) + exd_offset;

	auto text = e.code.substr(sxd, exd);

	e.code.erase(sxd, exd - sxd);
	RemoveTextEdit(e.code_str(), sxd, text.length());

	e.undo.push_back({operation_t{DELETE, std::string(text.c_str()), sxd, cursormove_t{e.row, e.col}}});
	e.__selection.CleanSelection();
//	e.UpdateLsp(false, string(e.code.Value()));
	set_update_parameters(true);
}

void Duplicate(void) {
	Focus();

	if (e.code.size() == 0) { return; }
	auto syd = LineOffset(e.code_str(), e.row) + 1;
	auto eyd = LineOffset(e.code_str(), e.row + 1) + 1;
	if (e.row == 0) { syd--; } // this is required for first line only
	auto duplicatedSlice = e.code.substr(syd, eyd);

	e.code.insert(eyd, duplicatedSlice);
	e.row++;

	eyd = LineOffset(e.code_str(), e.row) + 1;

	InsertTextEdit(e.code_str(), syd, duplicatedSlice.size());
	e.undo.push_back({operation_t{INSERT, std::string(duplicatedSlice.c_str()), eyd, cursormove_t{e.row, e.col}}});
//	e.UpdateLsp(false, string(e.code.Value()));
	set_update_parameters(true);
}

void OnUndo(void) {
	if (e.undo.empty()) { /*e.UpdateLsp(true, string(e.code.Value())); */return; }

	auto lastOperation = e.undo[e.undo.size() - 1];
	e.undo.pop_back(); // removing last element
	Focus();

	auto index = lastOperation.size() - 1;
	operation_t o{};
undo:
	if (lastOperation.size() > 0) { o = lastOperation[index]; } else { goto exit; }

	switch (o.action) {
	case INSERT:
	case ADDCHAR:
		e.code.erase(o.offset, o.offset - o.text.size());
		RemoveTextEdit(e.code_str(), o.offset, o.text.size());
		e.row = o.cursor.line; e.col = o.cursor.column;
	break;
	case DELETE:
		e.code.insert(o.offset, o.text.c_str());
		InsertTextEdit(e.code_str(), o.offset, o.text.size());
		e.row = o.cursor.line; e.col = o.cursor.column;
	break;
	default:
	break;
	}

	if (index > 0) { index--; goto undo; }
exit:
	e.redo.push_back(lastOperation);
//	e.UpdateLsp(false, string(e.code.Value()));
	set_update_parameters(true);
}

void OnRedo(void) {
	if (e.redo.size() == 0) { return; }

	auto lastRedoOperation = e.redo[e.redo.size() - 1];
	e.redo.pop_back(); // removing last element

	auto index = 0, offset = 0;
	operation_t o{};
redo:
	if (lastRedoOperation.size() > 0) { o = lastRedoOperation[index]; } else { goto exit; }

	switch (o.action) {
	case ADDCHAR:
		offset++;
	case INSERT:
		e.code.insert(o.offset, o.text.c_str());
		InsertTextEdit(e.code_str(), o.offset, o.text.size());
		e.row = o.cursor.line; e.col = o.cursor.column + offset;
	break;
	case DELETE:
		e.code.erase(o.offset, o.offset - o.text.size());
		RemoveTextEdit(e.code_str(), o.offset, o.text.size());
		e.row = o.cursor.line; e.col = o.cursor.column;
	break;
	default:
	break;
	}

	if (index < (int)lastRedoOperation.size() - 1) { index++; goto redo; }
exit:
	e.undo.push_back(lastRedoOperation);
//	e.UpdateLsp(false, string(e.code.Value()))
	set_update_parameters(true);
}
